// @bsv/overlay-discovery-services/dist/cjs/src/SHIP/SHIPTopicManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SHIPTopicManager = void 0;
const sdk_1 = require("@bsv/sdk");
const isTokenSignatureCorrectlyLinked_js_1 = require("../utils/isTokenSignatureCorrectlyLinked.js");
const isAdvertisableURI_js_1 = require("../utils/isAdvertisableURI.js");
const SHIPTopic_docs_js_1 = __importDefault(require("./SHIPTopic.docs.js"));
const isValidTopicOrServiceName_js_1 = require("../utils/isValidTopicOrServiceName.js");
/**
 * ðŸš¢ SHIP Topic Manager
 * Implements the TopicManager interface for SHIP (Service Host Interconnect Protocol) tokens.
 *
 * The SHIP Topic Manager identifies admissible outputs based on SHIP protocol requirements.
 * SHIP tokens facilitate the advertisement of nodes hosting specific topics within the overlay network.
 */
class SHIPTopicManager {
    /**
     * Identifies admissible outputs for SHIP tokens.
     * @param beef - The transaction data in BEEF format.
     * @param previousCoins - The previous coins to consider.
     * @returns A promise that resolves with the admittance instructions.
     */
    async identifyAdmissibleOutputs(beef, previousCoins) {
        const outputsToAdmit = [];
        try {
            const parsedTransaction = sdk_1.Transaction.fromBEEF(beef);
            for (const [i, output] of parsedTransaction.outputs.entries()) {
                try {
                    const result = sdk_1.PushDrop.decode(output.lockingScript);
                    if (result.fields.length !== 5)
                        continue; // SHIP tokens have 5 fields
                    const shipIdentifier = sdk_1.Utils.toUTF8(result.fields[0]);
                    if (shipIdentifier !== 'SHIP')
                        continue; // SHIP identifier must be present
                    const advertisedURI = sdk_1.Utils.toUTF8(result.fields[2]);
                    if (!(0, isAdvertisableURI_js_1.isAdvertisableURI)(advertisedURI))
                        continue; // Advertised URI must be acceptable
                    const topic = sdk_1.Utils.toUTF8(result.fields[3]);
                    if (!(0, isValidTopicOrServiceName_js_1.isValidTopicOrServiceName)(topic))
                        continue; // Topic or service name must be valid
                    if (!topic.startsWith('tm_'))
                        continue; // SHIP only accepts "tm_" (topic manager) advertisements
                    if (!(0, isTokenSignatureCorrectlyLinked_js_1.isTokenSignatureCorrectlyLinked)(result.lockingPublicKey, result.fields))
                        continue; // Signatures must be properly linked
                    outputsToAdmit.push(i);
                }
                catch (error) {
                    // It's common for other outputs to be invalid; no need to log an error here
                    continue;
                }
            }
        }
        catch (error) {
            // Only log an error if no outputs were admitted and no previous coins consumed
            if (outputsToAdmit.length === 0 && (!previousCoins || previousCoins.length === 0)) {
                console.error('â›´ï¸ Error identifying admissible outputs:', error);
            }
        }
        if (outputsToAdmit.length > 0) {
            console.log(`ðŸ›³ï¸ Ahoy! Admitted ${outputsToAdmit.length} SHIP ${outputsToAdmit.length === 1 ? 'output' : 'outputs'}!`);
        }
        if (previousCoins && previousCoins.length > 0) {
            console.log(`ðŸš¢ Consumed ${previousCoins.length} previous SHIP ${previousCoins.length === 1 ? 'coin' : 'coins'}!`);
        }
        if (outputsToAdmit.length === 0 && (!previousCoins || previousCoins.length === 0)) {
            console.warn('âš“ No SHIP outputs admitted and no previous SHIP coins consumed.');
        }
        return {
            outputsToAdmit,
            coinsToRetain: []
        };
    }
    /**
     * Returns documentation specific to the SHIP topic manager.
     * @returns A promise that resolves to the documentation string.
     */
    async getDocumentation() {
        return SHIPTopic_docs_js_1.default;
    }
    /**
     * Returns metadata associated with this topic manager.
     * @returns A promise that resolves to an object containing metadata.
     */
    async getMetaData() {
        return {
            name: 'SHIP Topic Manager',
            shortDescription: 'Manages SHIP tokens for service host interconnect.'
        };
    }
}
exports.SHIPTopicManager = SHIPTopicManager;
